response_operations_ui/templates/sign_in.html

```diff
@@ -4,72 +4,88 @@
 
 {% block main %}
 
-<div class="container page__container">
-    <div class="grid">
-        <div class="grid__col col-12@m">
-            <br>
+<div class="grid">
+    <div class="grid__col col-12@m">
+        <div role="main" class="page__main col-6@m">
             {% with messages = get_flashed_messages(with_categories=true) %}
                 {% if messages %}
                     {% for category, message in messages %}
                         {% if category == "failed_authentication" %}
-                        <div class="panel panel--error">
+                        <div class="panel panel--error u-mt-s">
                             <div class="panel__header">
                                 <h1 class="panel__title venus">{{ message }}</h1>
                             </div>
                             <div class="panel__body" data-qa="error-body">
-                                <p class="mars"><a href="#sign-in-details" id="try-again-link" class="js-inpagelink">Please try again</a></p>
+                                <p class="mars"><a href="#username" id="try-again-link" class="js-inpagelink">Please try again</a></p>
                             </div>
                         </div>
                         {% endif %}
                         {% if category == "successful_signout" %}
-                        <div class="panel panel--success">
+                        <div class="panel panel--simple panel--success u-mt-s">
                             <div class="panel__body" data-qa="success-body">
                                 <p class="mars" id="successfully-signed-out">{{ message }}</p>
                             </div>
                         </div>
                         {% endif %}
                     {% endfor %}
                 {% endif %}
-            {% endwith %}
-            <form method="post">
-                {{ form.csrf_token }}
-                <div>
-                    {% if form.errors %}
-                    <div class="panel panel--error">
-                        <div class="panel__header">
-                            {% if form.errors|length == 1 %}
-                            <h1 id="error-count" class="panel__title venus">There is 1 error</h1>
-                            {% else %}
-                            <h1 id="error-count" class="panel__title venus">There are {{ form.errors|length }} errors</h1>
-                            {% endif %}
+            <section>
+                <form method="post" class="form u-mt-m">
+                    {{ form.csrf_token }}
+                    <div>
+                        {% if form.errors %}
+                        <div class="panel panel--error u-mb-m">
+                            <div class="panel__header">
+                                {% if form.errors|length == 1 %}
```

Comment:
do not you already check form.errors in the previous statment ?

Reply:
yea, but I need to get the variable again to test the length

Reply:
you can have a function that return the right message in base of the errors lenght. So that you remove code duplication and IF statemnet that increase code complexity.

Reply:
Unless I'm missing something, calling form.errors after the first time is identical to assigning it to a variable and calling that variable multiple times.
https://github.com/wtforms/wtforms/blob/master/wtforms/form.py#L149 shows that it caches the result after it gets called once.
Assuming that's true, then why bother assigning a variable or subclassing or doing anything, when form.errors|length is readable and cheap to do twice.  Or am I missing something?

Reply:
easy, it s called code reuse and apply DRY principle . Keeps your code cleaner and tidy :)

Reply:
If it's already being cached (as the WTForms code suggests) then there's no reason to over-engineer here and just stick with being explicit.

Reply:
@jcox-dev is just because it s meant to be a clening task. Just because is a UI does not mean that is good to have few nested if and for condition with few code repetition. But mine is just a suggestion, but leave the code in the same state . Breaking DRY principle , Code complexity and generaly speaking we lack in UI strategy generating flask page that looks like the old JSP/ASP ;) BUt it s fine for me if you want leave like it is.

Reply:
I agree, we should probably have more partials that do one thing well or at least some improved documentation for presenting stuff on the front-end internally. Probably worth checking with tech leads to bring it up in their meetings.